### PR TITLE
Updated Modifications to obs setup code(setup*.f90) for ps/w/spd/q/t to add observation provider info into obs diagnostic file

### PR DESCRIPTION
**Pull request by GangZhao-NOAA**
Adding modifications to obs setup code(setup*.f90) for ps/q/t/w/spd, to include the observation provider and sub-provider information into the obsdiags filee. Adding new option "l_obsprvdiag" to trigger the output.   These provider information can be read with read_diag_conv and used in the AutoObsQC for 3DRTMA.

Also modifying the code in read_diag_conv.f90 under util/Analysis_Utilities/read_diag to read the provider information in the obs-diag file and write out in text format which can be used for auto-obsqc.

### CODE REVIEW

File: src/gsi/gsimod.F90 [hunk 1]
```diff
@@ -712,7 +719,8 @@ module gsimod
        write_fv3_incr,incvars_to_zero,incvars_zero_strat,incvars_efold,diag_version,&
        cao_check,lcalc_gfdl_cfrac,tau_fcst,efsoi_order,lupdqc,lqcoef,cnvw_option,l2rwthin,hurricane_radar,&
        npePgrp_rfv3,rfv3_pe_T,rfv3_pe_v,rfv3_pe_q,rfv3_pe_ps,rfv3_pe_dz, &
-       l_reg_update_hydro_delz
+       l_reg_update_hydro_delz, &
+       l_obsprvdiag
```
Comment (guoqing-noaa) [inline src/gsi/gsimod.F90]: Line 723 can be put into line 722.

File: src/gsi/gsimod.F90 [hunk 2]
```diff
@@ -1757,6 +1765,9 @@ subroutine gsimain_initialize
      end if
   end if
 
+! Ensure when running 2DVAR (RTMA) or 3DRTMA, writing out obs provider info
+  if (twodvar_regional .or. l_rtma3d) l_obsprvdiag = .true.
+
```
Comment (guoqing-noaa) [inline src/gsi/gsimod.F90]: It is better NOT to force (or hardcoded) this inside the code as it hides information from general users.
For 2D/3D RTMA, setting l_obsprvdiag = .true. in the namelist is enough and provides a flexibility/backward compatibility to NOT to output subprovider information.

File: src/gsi/setupps.f90 [hunk 3]
```diff
@@ -321,7 +328,10 @@ subroutine setupps(obsLL,odiagLL,lunin,mype,bwork,awork,nele,nobs,is,conv_diagsa
      ioff0=20
      nreal=ioff0
      if (lobsdiagsave) nreal=nreal+4*miter+1
-     if (twodvar_regional) then; nreal=nreal+2; allocate(cprvstg(nobs),csprvstg(nobs)); endif
+     if (twodvar_regional .or. l_rtma3d .or. l_obsprvdiag) then
```
Comment (guoqing-noaa) [inline src/gsi/setupps.f90]: Line 331 only needs `if (l_obsprvdiag) then`

As discussed before, let "l_obsprvdiag" solely decide what to do, not if it is 2D/3D RTMA.  If RTMA wants to output subprovider information, just set "l_obsprvdiag" to "true" in the namelist. 

Suggest to remove all these kind of unnecessary if (twodvar_regional .or. l_rtma3d in other places.

Also note that "l_rtma3d" is previously added for ghost reflectivity removal in the cloud analysis part. If a RTMA version does not run cloud analysis, using "if l_rtma3d" will be problematic.

Comment (GangZhao-NOAA) [reply src/gsi/setupps.f90]: I can remove the trigger of "l_rtma3d" in setup?.f90 to control the output of obs provider info. 
twodvar_regional might still be necessary for backward compatible to operational (2D)RTMA sono need to modify the script and namelist file to add the "l_obsprvdiag". Anyway, I will discuss with Manuel and Matthew for this.

File: src/gsi/gsimod.F90 (hunk 2 above)
Comment (GangZhao-NOAA) [reply src/gsi/gsimod.F90]: I will remove this line.

File: src/gsi/gsimod.F90 (hunk 1 above)
Comment (GangZhao-NOAA) [reply src/gsi/gsimod.F90]: I will fix it.

**Comment by GangZhao-NOAA** [comment]
@guoqing-noaa  @hu5970 

Thank you very much for your comments/suggestions on the code. I will discuss with Manuel and Matthew about the necessity of keeping "twodvar_regional" in setup??.f90. 
Today aftger discussion with Matthew, I found a few questions in the read_diag_conv.f90. Matthew is actually running the code without using namelist, but reading the parameters from command line (like read_diag_conv.x infilename outfilename true false). I am asking them which way (namelist or command argument list) they prefer to run.
I may have to take the PR back for update, then re-submit for review.

Thank you very much!

**Comment by guoqing-noaa** [comment]
@GangZhao-NOAA Thanks for looking on this. You don't need to close this PR and then create a new one. Just push new changes to your feature/rrfs branch and it will be reflected here.

File: src/gsi/setupq.f90 [hunk 4]
```diff
@@ -394,7 +400,11 @@ subroutine setupq(obsLL,odiagLL,lunin,mype,bwork,awork,nele,nobs,is,conv_diagsav
      ioff0=21
      nreal=ioff0
      if (lobsdiagsave) nreal=nreal+4*miter+1
-     if (twodvar_regional) then; nreal=nreal+2; allocate(cprvstg(nobs),csprvstg(nobs)); endif
+     if (twodvar_regional .or. l_obsprvdiag) then
+       nreal=nreal+2                            ! account for idomsfc,izz
+       allocate(cprvstg(nobs),csprvstg(nobs))   ! obs provider info
+       if(l_pbl_pseudo_surfobsq) allocate(cprvstgp(nobs*3),csprvstgp(nobs*3)) ! obs provider info for pseudo obs
+       endif
```
Comment (guoqing-noaa) [inline src/gsi/setupq.f90]: remove two spaces before line 407 so that it aligns with line 403

**Review by guoqing-noaa — APPROVED** [review]
Looks good to me (with one minor formatting comment). Thanks!

Another thing, have you run comparison tests to make sure this PR dose not change the GSI analysis result?

File: src/gsi/setupq.f90 (hunk 4 above)
Comment (GangZhao-NOAA) [reply src/gsi/setupq.f90]: @guoqing-noaa @hu5970
Thank you for the review. 
The two spaces in beginning of the line 407 of "setupq/f90" was removed. I make a new commit and push it back. 

As for comparing with old code for validation, I have not run it yet.  My colleague Matthew Morris is going to test it. I will let you know the results of comparison. But he will only test with HRRR-based run, not for FV3-LAM based rrfs run.

Thank you!

**Comment by GangZhao-NOAA** [comment]
@guoqing-noaa @hu5970 
Hi, Ming and Guoqing,

I found an issue with the "read_diag_conv.f90" and need to discuss with Manuel and Matthew about the way to fix it. Please do not merge the code for now. 
Thank you very much! Happy Holiday!

- Gang

**Comment by GangZhao-NOAA** [comment]
@hu5970 @guoqing-noaa 
Hi Guoqing and Ming,

I fixed the bugs in read_diag_conv.f90 (see the latest [commit # 295fdf](https://github.com/NOAA-GSL/GSI/pull/8/commits/e295fdf412a814f9f7c0ca0639e865dc300ab4a3)). Matthew Morris of EMC had tested it and confirm that it is correct now.

As for your request to compare the analysis results from my modified GSI code (feature/rrfs in my fork of GSI) and the old GSI code (feature/rrfs of NOAA-GSL GSI), I compiled both GSI package on WCOSS. Matthew made the test runs on WCOSS. The analysis results are identical. So I believe the modified code are good.

So the modified code should be ready for merging into NOAA-GSL/GSI. We hope the merging could be done soon, so we can work on the same code.

Thank you very much!

- Gang

**Comment by GangZhao-NOAA** [comment]
@guoqing-noaa @hu5970 
Thank you so much!

**Pull request merged**